node/src/stratum.rs

```diff
@@ -0,0 +1,1637 @@
+        let version_rolling_mask = config_map.get("version-rolling.mask").or(None);
+        let version_rolling_min_bit_count =
+            config_map.get("version-rolling.min-bit-count").or(None);
+        if version_rolling_mask.is_none() == false {
+            let mut mask_bytes: [u8; 4] = [0u8; 4];
+            let version_rolling_mask_str = version_rolling_mask.unwrap().as_str().unwrap();
+            hex::decode_to_slice(version_rolling_mask_str, &mut mask_bytes);
+
+            let final_rollable_version_bits = u32::from_be_bytes(mask_bytes) & 0x1FFFE000;
+            // `0x1FFFE000` is a reasonable default as it allows all 16 version bits to be used
+            let hex_str = u32::to_string(&final_rollable_version_bits);
+            self.version_rolling_mask = Some(hex_str);
+        }
+        if version_rolling_min_bit_count.is_none() == false {
+            let mut mask_bytes: [u8; 4] = [0u8; 4];
+            let version_rolling_min_bit_count_str =
+                version_rolling_min_bit_count.unwrap().as_str().unwrap();
+            hex::decode_to_slice(version_rolling_min_bit_count_str, &mut mask_bytes);
+            self.version_rolling_min_bit = Some(u32::from_be_bytes(mask_bytes));
+        }
+
+        Ok(StratumResponses::StandardResponse {
+            std_response: StandardResponse {
+                id: Some(client_request_id),
+                result: Some(json!({
+                    "minimum-difficulty":false,
+                    "version-rolling": true,
+                    "version-rolling.mask":self.version_rolling_mask.clone().unwrap_or("1fffe000".to_string()),
+                    "version-rolling.min-bit-count":self.version_rolling_min_bit.unwrap_or(0)
+
+                })),
+                error: None,
+            },
+        })
+    }
+    ///The optional second parameter specifies a mining.notify subscription id the client wishes to resume working with (possibly due to a dropped connection). If provided, a server MAY (at its option) issue the connection the same extranonce1. Note that the extranonce1 may be the same (allowing a resumed connection) even if the subscription id is changed!
+    /// The result contains three items:
+
+    /// Subscriptions. - An array of 2-item tuples, each with a subscription type and id.
+    /// ExtraNonce1. - Hex-encoded, per-connection unique string which will be used for creating generation transactions later.
+    /// ExtraNonce2_size. - The number of bytes that the miner users for its ExtraNonce2 counter.
+    pub async fn handle_subscribe(
+        &mut self,
+        subscribe_req_params: &Value,
+        client_request_id: u64,
+    ) -> Result<StratumResponses, StratumErrors> {
+        log::info!("Subscribing is taking place -- {:?}", subscribe_req_params);
+        //FIXME dummy testing subscription IDs must be unique though
+        let subscriptions: Vec<(String, String)> = vec![
+            (String::from("mining.set_difficulty"), String::from("34")),
```

Comment:
Can you explain these constants "34" and "12"? I think one of them should be a UUID, no?